pkg/ignore/dockerignore.go

```diff
@@ -0,0 +1,167 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ignore
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+const (
+
+	// dockerIgnoreFile defines the Name of the scanner file
+	dockerIgnoreFile = ".dockerignore"
+	// invertPrefix is used to prefix the Patterns to symbolize inversions
+	invertPrefix = "!"
+)
+
+var (
+	_ FileIgnorer = &defaultIgnorer{}
+	// root directory to start the scanner
+	directory string
+	// patterns holds all the possible Ignorable patterns
+	ignorePatterns []fileIgnorePattern
+)
+
+// dockerIgnorer processes .dockerignore and use the extracted Patterns for ignoring files
+type dockerIgnorer struct {
+}
+
+// NewOrDefault builds and returns the new or default FileIgnorer interface
+// In all file not found cases the this wil return the default ignorer
+func NewOrDefault(dir string) (FileIgnorer, error) {
+
+	// every new gets clean includes
+	ignorePatterns = make([]fileIgnorePattern, 0)
+
+	directory = dir
+
+	ignoreFile := filepath.Join(dir, dockerIgnoreFile)
+
+	p, err := scanAndBuildPatternsList(ignoreFile)
+
+	rawPatterns = append(rawPatterns, p...)
+
+	switch {
+	case os.IsNotExist(err):
+		return &defaultIgnorer{}, nil
+	case err != nil:
+		return nil, err
+	}
+
+	for _, ip := range p {
+		x, err := toFileIgnorePattern(ip)
+		if err != nil {
+			return nil, err
+		}
+		ignorePatterns = append(ignorePatterns, *x)
+	}
+
+	return &dockerIgnorer{}, nil
+}
+
+// scanAndBuildPatternsList takes the file typically the .dockerignore file
+// splits the file by new line (\n) and normalize them with following rules
+// - removes the UTF8 Byte Order Mark (BOM) characters
+// - scanner comments line starting with "#"
+// - trim spaces in the pattern
+// - make the pattern as clean filenames using golang filepath utils
+func scanAndBuildPatternsList(ignoreFile string) ([]string, error) {
+
+	var patterns []string
+	patterns = append(patterns, defaultPatterns...)
+
+	fr, err := os.Open(ignoreFile)
+
+	switch {
+	case os.IsNotExist(err):
+		return patterns, nil
+	case err != nil:
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(fr)
+
+	// UTF8 byte order mark (BOM) which are typically first three bytes of the file with
+	// the hexadecimal characters: EF,BB,BF
+	utf8bom := []byte{0xEF, 0xBB, 0xBF}
+	lineNo := 0
+
+	for scanner.Scan() {
+		scannedBytes := scanner.Bytes()
+
+		if lineNo == 0 {
+			scannedBytes = bytes.TrimPrefix(scannedBytes, utf8bom)
+		}
+
+		pattern := string(scannedBytes)
```

Comment:
I think the body of this function from roughly here down could be encapsulated into a small function `string -> string` that we could extensively table test.